[willowisp01] iP

willowisp01 opened the pull request:
# Luke

> Like Duke, but starting with L instead. -- (Myself)

Luke is like Duke but better. That's because you can
- add different types of events
- type commands, like, really fast
- form a unique bond with Luke

You can download it [here](https://github.com/willowisp01/ip/releases/tag/A-Jar) for the price of **zero dollars and zero cents**!

Features: 

- [x] Managing Tasks 
- [x] Managing Deadlines
- [x] Tsundere Personality 💢💢💢 👉👈
- [ ] Other Personalities (WIP)

The full source code:
```
public static void main(String[] args) {
   while (true) {
      respondToInput(input)
   }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

willowisp01 committed:
add greeting and goodbye

willowisp01 committed:
echo commands, close when given command "bye"

willowisp01 committed:
add tasklist

willowisp01 committed:
add new Task class

willowisp01 committed:
can mark tasks as done

willowisp01 committed:
add method fullstats to display full info of task

willowisp01 committed:
implement todo class

willowisp01 committed:
implement deadline class

willowisp01 committed:
implement event class

willowisp01 committed:
add task selection logic

willowisp01 committed:
fix whitespace and syntax issues

willowisp01 committed:
implement regression testing

willowisp01 committed:
add exception handling

willowisp01 committed:
add delete functionality

willowisp01 committed:
Split logic in main() into different methods for separation of concerns

willowisp01 committed:
move more logic out of main() and into other methods

willowisp01 committed:
Fix filepath issues

willowisp01 committed:
make separate class for history (prepare for serialization)

willowisp01 committed:
shift more logic out of main into history

willowisp01 committed:
Implement Serializable

willowisp01 committed:
Complete history feature

willowisp01 committed:
Reset save data

willowisp01 committed:
Merge branch 'branch-Level-7'

willowisp01 committed:
Add datetime support

willowisp01 committed:
Add some testing files

willowisp01 committed:
Merge branch 'branch-Level-8'

willowisp01 committed:
Update gitignore

willowisp01 committed:
Split logic into Parser, History, Tasklist classes

willowisp01 committed:
fix history

willowisp01 committed:
Add simple UI class

willowisp01 committed:
Rename History class to Storage

willowisp01 committed:
Run testcases

willowisp01 committed:
Merge branch 'A-MoreOOP'

willowisp01 committed:
Reset to working version

willowisp01 committed:
Move classes to package luke

willowisp01 committed:
Move classes to package luke

willowisp01 committed:
Merge branch 'A-Packages'

willowisp01 committed:
Set gradle run config

willowisp01 committed:
Add tests for Parser, TaskList, and Todo

willowisp01 committed:
Merge branch 'A-JUnit'

willowisp01 committed:
Create jar file

willowisp01 committed:
Merge branch 'A-Jar'

willowisp01 committed:
Create javadocs

willowisp01 committed:
Remove unnecessary files

willowisp01 committed:
Merge branch 'branch-A-JavaDoc'

willowisp01 committed:
Udpate code to match coding standard

willowisp01 committed:
Merge branch 'branch-A-CodingStandard'

willowisp01 committed:
Add functionality to find tasks

willowisp01 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/luke/Parser.java
#	src/main/java/luke/TaskList.java

Alteqa commented on src/main/java/luke/TaskList.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    public void findTask(String keyword) {
+        int num = 1;
+        boolean found = false;
```
Should be isFound or hasBeenFound instead of simply just found.

Alteqa commented on src/main/java/luke/UI.java:
```diff
@@ -0,0 +1,26 @@
+package luke;
+
+public class UI {
+    private static String logo = "  _____             __             \n"
```
constants should be all capitalised such as LOGO instead of just logo.

Alteqa commented on src/main/java/luke/Luke.java:
```diff
@@ -0,0 +1,78 @@
+public class Luke {
+
+    //Logo created using https://patorjk.com/software/taag/#p=display&f=Varsity&t=Luke
+    private static String logo = "  _____             __             \n"
```
Constants should be capitalised.

Alteqa commented on src/main/java/luke/TaskList.java:
```diff
@@ -0,0 +1,112 @@
+    public void addTask(Task task) {
+        if (task != null) {
+            tasks.add(task);
+            System.out.println("I helped you add task '" + task.fullStatus() + "'. But do it yourself next time! Hmmph!" + "\n");
```
code should be wrapped/split by the + instead of the string itself.

Alteqa commented on src/main/java/luke/Luke.java:
```diff
@@ -0,0 +1,78 @@
+        UI ui = new UI();
+        ui.greet();
+
+        String wd = System.getProperty("user.dir");
```
variable names should be descriptive.

Austintjh19 replied:
The process of initialising the storage file may be more suitable to be done by the Storage class.

Alteqa reviewed (commented):
Overall your code was excellent and there were very few errors. Even with errors that I found they were very minor and did not affect my understanding of your programme at all.

Austintjh19 commented on src/main/java/luke/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.time.LocalDate;
+
+public class Deadline extends Task {
```
Maybe "DeadlineTask" as a class name might be more relevant and descriptive compared to simply "Deadline." Might have a Deadline task in the future. Applicable to Event & Todo as well. :)

Austintjh19 commented on src/main/java/luke/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+import java.time.LocalDate;
+
+public class Deadline extends Task {
+    private LocalDate by;
```
byDate may be more applicable as a variable name. Given that you are storing it as a LocalDate. Future programmers will know it doesn't store for instance Tuesday or 3 pm.  Applicable to Event as well.

Austintjh19 commented on src/main/java/luke/Luke.java:
```diff
@@ -0,0 +1,78 @@
+
+        //load the file if there is save data (reference: https://www.baeldung.com/java-serialization)
+        //otherwise, create a new History object.
+        File historyFile = new File(String.valueOf(historyPath));
```
The process of loading the storage file may be more suitable to be done by the Storage class.

Austintjh19 commented on src/main/java/luke/Parser.java:
```diff
@@ -0,0 +1,142 @@
+     */
+    public boolean parseCommand(String input) {
+        String trimmedLowercase = input.trim().toLowerCase();
+        if (trimmedLowercase.equals("bye")) {
```
Maybe consider switching to a case statement rather than relying on if-else constructs, it could potentially enhance code readability.

Austintjh19 commented on src/main/java/luke/Storage.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.ArrayList;
+import java.io.IOException;
+
+public class Storage implements Serializable {
```
Considering future iterations of Luke, it may be beneficial to create a separate class, such as TxtStorage, that inherits from this file. This approach allows for potential expansion beyond storing data solely in text files in future versions of the application. Just a suggestion.

Austintjh19 commented on src/main/java/luke/Storage.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Storage implements Serializable {
+
+    private ArrayList<Task> history;
```
Maybe a more appropriate name ? e.g. TasksHistory

Austintjh19 commented on src/main/java/luke/Task.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Marks this task as complete.
+     */
+    public void complete() {
```
Maybe setIsDone is more appropriate ?

Austintjh19 commented on src/main/java/luke/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+     * @return The status of task. ([taskType][isDone] taskname by)
+     */
+    @Override
+    public String fullStatus() { //TODO: add type of task later
```
I think using a verb as part of the function name is more appropriate. E.g. getFullStatus ?

Austintjh19 commented on src/main/java/luke/TaskList.java:
```diff
@@ -0,0 +1,112 @@
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("Jeez, you really ought to give me a number I can work with... got that?");
+            System.out.println("[Item index exceeds history count]\n");
+            return;
```
Maybe return a boolean. True/ false representing whether the task was successfully executed. Instead of having an empty return.

Austintjh19 reviewed (commented):
Good work. really well written code. Just a few suggestion here and there from me.

willowisp01 committed:
Update README.md

willowisp01 committed:
Get javafx imports to work

willowisp01 committed:
Make multiple methods return string instead of print to console (adapted from https://se-education.org/guides/tutorials/javaFxPart2.html)

willowisp01 committed:
Handle not a number exceptions

willowisp01 committed:
Update frame to fit in center(ish). Update user and luke image.

willowisp01 committed:
Merge branch 'master' of https://github.com/willowisp01/ip into Level-10

willowisp01 committed:
Credit sources

willowisp01 committed:
Update application image

willowisp01 committed:
Create new jar

willowisp01 committed:
Rename variables and methods according to coding standard

Rename methods to make it immediately obvious what it should do.
Rename constants such as LOGO to be in caps.

willowisp01 committed:
Split the role of Parser

Created a new class TaskMaker to make tasks as currently Parser is both
making Tasks and Parsing them. Each class should ideally do its own
job.

willowisp01 committed:
Make Parser neater according to SLAP

Created more layers of abstraction through the use of helper
functions and enums. This makes what Parser is doing a lot clearer.

willowisp01 committed:
Move loading a save logic from Luke class to Storage class

willowisp01 committed:
Merge pull request #1 from willowisp01/branch-A-CodeQuality

Branch a code quality

willowisp01 committed:
Create assertions

One important assertion is that the Storage should never be null.
This is because it should either be loaded or created.

willowisp01 committed:
Merge pull request #3 from willowisp01/branch-A-Assertions

Create assertions

willowisp01 committed:
Add reminder feature (use command remind)

willowisp01 committed:
Merge pull request #4 from willowisp01/branch-BCD-Extension

Add reminder feature (use command remind)

willowisp01 committed:
Merge branch 'branch-BCD-Extension'

willowisp01 committed:
Merge branch 'master' of https://github.com/willowisp01/ip

willowisp01 committed:
Fix test cases

willowisp01 committed:
Merge branch 'fix-test-cases'

willowisp01 committed:
Update Javadocs

willowisp01 committed:
Fixed miscellaneous issues given by soc-se-script

willowisp01 committed:
Credit image sources

willowisp01 committed:
Add UI.png

willowisp01 committed:
Update readme

willowisp01 committed:
Merge branch 'finalize-code'

willowisp01 committed:
Move readmes to docs

willowisp01 committed:
Update README.md

willowisp01 committed:
Update README

willowisp01 committed:
Update build file to target java 11
